TrickfireRobotics/NasaRmc2018
# Light Detection

# Description

I wrote a node that detects sharp changes in light. The detection action server. 

It is started by giving it a goal, where it spins up and starts looking at a configurable camera feed. It keeps a moving average of color intensities (0->255) for each color channel. Currently when it detects a significant image intensity change it marks it's goal to complete as a signal to the action client and stops image processing. 

Some configurable paramters I left exposed to allow us to tweak sensitivity are:
  - average width: How long of a moving average
  - intensity threshold: How sharp of a spike is considered turning on the light.

Now because we don't have the led for this set up yet, and don't know what color it is. I don't have this set to detect any particular color, instead it just weighs them easily. This is a one line of code fix when we have the device so I am not particularly worried.

However the results are pretty good, at least on my webcam. So in theory they will get even more robust with the actual light.

# Testing
I wrote a test class `light_detection_action_client` to test this code, and show off it's functionality. I made use of the `cv_camera` to generate the images. In my case it found my laptop webcam out of the box. Let me know if you have any trouble getting an image, and I can help you troubleshoot.

The launch file can be run with `roslaunch tfr_launch light_test.launch`.

After running you should see the action server:
1. start
2. connect with the client
3. accept a goal
4. get preempted 
5. start a new goal
6. wait for a light

At this point you need to mimic a light being turned on, either turn on your phone flashlight next to the camera aperture, or cover and uncover it with your palm. 

7. a success message printed

There isn't much math we want to test here that's any more complicated than an average, all of the heavy lifting is done by the opencv matrix libraries.
If you want to see calculated intensities, uncomment line 74 of light_detector.cpp.

__EDIT:__
So there was an extra commit that was on here that sets it up on the jetson camera I added an extra launch file to test at home. Do:  `roslaunch tfr_launch local_light_test.launch`.
